samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java

```diff
@@ -20,28 +20,70 @@
 
 import org.apache.samza.config.Config;
 import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.functions.PartialJoinFunction;
+import org.apache.samza.operators.functions.PartialJoinFunction.PartialJoinMessage;
 import org.apache.samza.operators.spec.PartialJoinOperatorSpec;
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.storage.kv.KeyValueIterator;
+import org.apache.samza.storage.kv.KeyValueStore;
 import org.apache.samza.task.MessageCollector;
 import org.apache.samza.task.TaskContext;
 import org.apache.samza.task.TaskCoordinator;
 
+import java.util.ArrayList;
+import java.util.List;
 
 /**
- * Implementation of a {@link PartialJoinOperatorSpec}. This class implements function
- * that only takes in one input stream among all inputs to the join and generate the join output.
+ * Implementation of a {@link PartialJoinOperatorSpec} that joins messages of type {@code M} in this stream
+ * with buffered messages of type {@code JM} in the other stream.
  *
  * @param <M>  type of messages in the input stream
  * @param <JM>  type of messages in the stream to join with
  * @param <RM>  type of messages in the joined stream
  */
-class PartialJoinOperatorImpl<M, K, JM, RM> extends OperatorImpl<M, RM> {
+class PartialJoinOperatorImpl<K, M, JM, RM> extends OperatorImpl<M, RM> {
 
-  PartialJoinOperatorImpl(PartialJoinOperatorSpec<M, K, JM, RM> joinOp, MessageStreamImpl<M> source, Config config, TaskContext context) {
-    // TODO: implement PartialJoinOperatorImpl constructor
+  private final PartialJoinFunction<K, M, JM, RM> thisPartialJoinFn;
+  private final PartialJoinFunction<K, JM, M, RM> otherPartialJoinFn;
+  private final long ttlMs;
+
+  PartialJoinOperatorImpl(PartialJoinOperatorSpec<K, M, JM, RM> partialJoinOperatorSpec, MessageStreamImpl<M> source,
+      Config config, TaskContext context) {
+    this.thisPartialJoinFn = partialJoinOperatorSpec.getThisPartialJoinFn();
+    this.otherPartialJoinFn = partialJoinOperatorSpec.getOtherPartialJoinFn();
+    this.ttlMs = partialJoinOperatorSpec.getTtlMs();
   }
 
   @Override
   public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
-    // TODO: implement PartialJoinOperatorImpl processing logic
+    K key = thisPartialJoinFn.getKey(message);
+    thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, System.currentTimeMillis()));
+    PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
+    if (otherMessage != null) {
+      RM joinResult = thisPartialJoinFn.apply(message, otherMessage.getMessage());
+      this.propagateResult(joinResult, collector, coordinator);
+    }
+  }
+
+  @Override
+  public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    KeyValueStore<K, PartialJoinMessage<M>> thisState = thisPartialJoinFn.getState();
+    KeyValueIterator<K, PartialJoinMessage<M>> iterator = thisState.all();
```

Comment:
It would also be a nice practice to:
1) bound the time to spend on this GC work to avoid long pause
2) close the iterator after the batch delete (since RocksDB has poor perf when iterators are left open)